src/bindings/js/node/src/infer_request.cpp

```diff
@@ -190,3 +197,82 @@ void InferRequestWrap::infer(const Napi::Object& inputs) {
 Napi::Value InferRequestWrap::get_compiled_model(const Napi::CallbackInfo& info) {
     return CompiledModelWrap::Wrap(info.Env(), _infer_request.get_compiled_model());
 }
+
+struct TsfnContext {
+    TsfnContext(Napi::Env env) : deferred(Napi::Promise::Deferred::New(env)){};
+
+    Napi::Promise::Deferred deferred;
+    ov::InferRequest* _context_ir;
+    std::thread nativeThread;
+
+    Napi::ThreadSafeFunction tsfn;
+    std::vector<ov::Tensor> input_tensors;
+    std::map<std::string, ov::Tensor> result;
+};
+
+void FinalizerCallback(Napi::Env env, void* finalizeData, TsfnContext* context) {
+    context->nativeThread.join();
+    delete context;
+};
+
+void performInferenceThread(TsfnContext* data) {
+    infer_mutex.lock();
+    for (size_t i = 0; i < data->input_tensors.size(); ++i) {
+        data->_context_ir->set_input_tensor(i, data->input_tensors[i]);
+    }
+    data->_context_ir->infer();
+
+    auto compiled_model = data->_context_ir->get_compiled_model().outputs();
+    std::map<std::string, ov::Tensor> outputs;
+
+    for (auto& node : compiled_model) {
+        auto tensor = data->_context_ir->get_tensor(node);
+        auto new_tensor = ov::Tensor(tensor.get_element_type(), tensor.get_shape());
+        tensor.copy_to(new_tensor);
+        outputs.insert({node.get_any_name(), new_tensor});
+    }
+
+    data->result = outputs;
+    infer_mutex.unlock();
+
+    auto callback = [](Napi::Env env, Napi::Function _, TsfnContext* data) {
+        auto m = data->result;
+        auto outputs_obj = Napi::Object::New(env);
+
+        for (const auto& [key, tensor] : m) {
+            auto new_tensor = ov::Tensor(tensor.get_element_type(), tensor.get_shape());
+            tensor.copy_to(new_tensor);
+            outputs_obj.Set(key, TensorWrap::Wrap(env, new_tensor));
+        }
+        data->deferred.Resolve({outputs_obj});
+    };
```

Comment:
We have to copy tensors from inferRequest output before we release the mutex (because they may be overwritten by other inferAsync() calls) 
The 'callback' loop sets output tensors as properties of Napi::Object `outputs_obj`. This object is created here (just before returning to js loop) to make sure it will not be garbage collected or lose some references to data.